src/main/kotlin/org/digma/intellij/plugin/ui/tests/TestsServiceImpl.kt

```diff
@@ -0,0 +1,163 @@
+    private var fillerOfLatestTests: FillerOfLatestTests? = null
+
+    init {
+
+        project.messageBus.connect(this).subscribe(EnvironmentChanged.ENVIRONMENT_CHANGED_TOPIC, object : EnvironmentChanged {
+            override fun environmentChanged(newEnv: String?, refreshInsightsView: Boolean) {
+                // dp nothing
+            }
+
+            override fun environmentsListChanged(newEnvironments: MutableList<String>?) {
+                sendOperativeEnvironments()
+            }
+        })
+
+    }
+
+    override fun dispose() {
+        //nothing to do
+    }
+
+    override fun initWith(cefBrowser: CefBrowser, fillerOfLatestTests: FillerOfLatestTests) {
+        this.cefBrowser = cefBrowser
+        this.fillerOfLatestTests = fillerOfLatestTests
+    }
+
+    private fun scope(): Scope {
+        return project.service<InsightsModelReactHolder>().model.scope
+    }
+
+    override fun refresh() {
+        Backgroundable.ensurePooledThread {
+            val scopeRequest = getScopeRequest()
+            scopeRequest?.let {
+                fillerOfLatestTests!!.fillDataOfTests(cefBrowser!!, it)
+            }
+        }
+    }
+
+    override fun getScopeRequest(): ScopeRequest? {
+        val scope = scope()
+
+        val spans: MutableSet<String> = mutableSetOf()
+        var methodCodeObjectId: String? = null
+        var endpointCodeObjectId: String? = null
+
+        when (scope) {
+            is MethodScope -> {
+                val methodInfo = scope.getMethodInfo()
+                methodCodeObjectId = methodInfo.idWithType()
+                if (methodInfo.hasRelatedCodeObjectIds()) {
+                    spans.addAll(methodInfo.spans.map { it.idWithType() })
+
+                    endpointCodeObjectId = methodInfo.endpoints.firstOrNull()?.idWithType()
+                }
+            }
+
+            is CodeLessSpanScope -> {
+                spans.add(CodeObjectsUtil.addSpanTypeToId(scope.getSpan().spanId))
+            }
+
+            is EndpointScope -> {
+                endpointCodeObjectId = CodeObjectsUtil.addEndpointTypeToId(scope.getEndpoint().id)
+            }
+
+            else -> {
+                return null
+            }
+        }
+
+        val scopeRequest = ScopeRequest(spans, methodCodeObjectId, endpointCodeObjectId)
+        println("DBG: scopeRequest=$scopeRequest")
+        return scopeRequest
+    }
+
+    // return JSON as string (type SetEnvironmentsMessage)
+    override fun sendOperativeEnvironments() {
+        if (cefBrowser == null) return
+
+        val envs = getOperativeEnvironmentEntities()
+
+        val setEnvironmentsMessage = SetEnvironmentsMessage(
+            JCefMessagesUtils.REQUEST_MESSAGE_TYPE,
+            JCefMessagesUtils.GLOBAL_SET_ENVIRONMENTS,
+            SetEnvironmentsMessagePayload(
+                envs
+            )
+        )
+
+        serializeAndExecuteWindowPostMessageJavaScript(cefBrowser!!, setEnvironmentsMessage)
```

Comment:
possible NPE.  although there is a test above , because cefBrowser is var it may be changed during this time. just kotlin best practices , use cefBrowser?.let{}

Reply:
look at line 116, I'm verifying its not null at the beginning of method

Reply:
I wrote above,theoretically a nuallable var may be changed from the time you check if its null until the time you use it.  in another thread for example. that's why best practice is to use ?.let

Reply:
in this case its irrelevant, its being set ONCE only